python/tutorials/02-fused-softmax.py

```diff
@@ -71,58 +72,85 @@ def naive_softmax(x):
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    input_ptrs = row_start_ptr + col_offsets
-    # Load the row into SRAM, using a mask since BLOCK_SIZE may be > than n_cols
-    row = tl.load(input_ptrs, mask=col_offsets < n_cols, other=-float('inf'))
-    # Subtract maximum for numerical stability
-    row_minus_max = row - tl.max(row, axis=0)
-    # Note that exponentiation in Triton is fast but approximate (i.e., think __expf in CUDA)
-    numerator = tl.exp(row_minus_max)
-    denominator = tl.sum(numerator, axis=0)
-    softmax_output = numerator / denominator
-    # Write back output to DRAM
-    output_row_start_ptr = output_ptr + row_idx * output_row_stride
-    output_ptrs = output_row_start_ptr + col_offsets
-    tl.store(output_ptrs, softmax_output, mask=col_offsets < n_cols)
+def softmax_kernel(output_ptr, input_ptr, input_row_stride, output_row_stride, n_rows, n_cols, XBLOCK: tl.constexpr,
+                   RBLOCK: tl.constexpr, NUM_REMAINDER_ROWS: tl.constexpr, num_stages: tl.constexpr):
+    pid = tl.program_id(0)
+    # starting row of the program
+    row_start = pid * (XBLOCK - 1) + min(pid, NUM_REMAINDER_ROWS)
+
+    for i in tl.range(0, XBLOCK, num_stages=num_stages):
+        row_idx = row_start + i
+        # The stride represents how much we need to increase the pointer to advance 1 row
+        row_start_ptr = input_ptr + row_idx * input_row_stride
+        # The block size is the next power of two greater than n_cols, so we can fit each
+        # row in a single block
+        col_offsets = tl.arange(0, RBLOCK)
+        input_ptrs = row_start_ptr + col_offsets
+        # Load the row into SRAM, using a mask since BLOCK_SIZE may be > than n_cols
+        mask = (i < XBLOCK - 1 or pid < NUM_REMAINDER_ROWS) and col_offsets < n_cols
+        row = tl.load(input_ptrs, mask=mask, other=-float('inf'))
+        # Subtract maximum for numerical stability
+        row_minus_max = row - tl.max(row, axis=0)
+        # Note that exponentiation in Triton is fast but approximate (i.e., think __expf in CUDA)
+        numerator = tl.exp(row_minus_max)
+        denominator = tl.sum(numerator, axis=0)
+        softmax_output = numerator / denominator
+        # Write back output to DRAM
+        output_row_start_ptr = output_ptr + row_idx * output_row_stride
+        output_ptrs = output_row_start_ptr + col_offsets
+        tl.store(output_ptrs, softmax_output, mask=mask)
 
 
 # %%
 # We can create a helper function that enqueues the kernel and its (meta-)arguments for any given input tensor.
 
+device = torch.cuda.current_device()
+NUM_SM = driver.active.utils.get_device_properties(device)["multiprocessor_count"]
+
 
 def softmax(x):
     n_rows, n_cols = x.shape
-    # The block size is the smallest power of two greater than the number of columns in `x`
-    BLOCK_SIZE = triton.next_power_of_2(n_cols)
+    num_programs = NUM_SM
+
+    # Create a number of persistent programs as many as SMs.
+    grid = lambda meta: (num_programs, )
+
```

Comment:
The gpu (A100, H100) can run 64 warps concurrently per SM, so I suspect we'd get better perf with `grid = (NUM_SM * 64 // num_warps,)`

Reply:
I think what you want to do here is to get the register / shared mem usage of the kernel and compute the occupancy . Then you can set `NUM_SM * occupancy` for the grid size.

Reply:
Thanks for the suggestion. I tried with `triton.compile` to compile the kernel ahead of the time, but I could not get `n_regs` before actually launching the kernel by when `driver.load_binary` is called. Do you think that can be worked around?

Reply:
So I figured out a way to get the kernel attributes such as n_regs by using `Triton.compile`, but it appears that the AOT compilation for each shape slows down the end-to-end performance in general. I tried to do AOT compile only once with a given shape and use that kernel to guide the computation of the grid size for all shapes. It gave good results but I feel that the kernel may not be representative for all shapes.